### PR TITLE
feat: allow the customization of analyzer rules

### PR DESCRIPTION
**Pull request by waynexia**
# Which issue does this PR close?

Closes #.

# Rationale for this change

Like `Optimizer` allows to customize its rules, this patch adds a similar interface to `Analyzer` and makes it public.

# What changes are included in this PR?

Allow users to insert their customized analyzer rules into the optimizer.

# Are these changes tested?

# Are there any user-facing changes?

Yes but not breaking. This publics the `Analyzer` and `AnalyzerRule`. And add corresponding config interfaces to `SessionContext`

### CODE REVIEW

File: datafusion/optimizer/src/optimizer.rs [hunk 1]
```diff
@@ -244,12 +256,18 @@ impl Optimizer {
...
-        Self { rules }
+    pub fn with_rules(
+        analyzer_rules: Vec<Arc<dyn AnalyzerRule + Send + Sync>>,
+        rules: Vec<Arc<dyn OptimizerRule + Send + Sync>>,
+    ) -> Self {
+        Self {
+            analyzer_rules,
+            rules,
+        }
     }
```
Comment (waynexia) [inline datafusion/optimizer/src/optimizer.rs:271]: I'm unsure if this is a proper place to customize analyzer rules. But from the code structure `Analyzer` is part of `Optimizer`. And only `Optimizer` is exposed to the context. Please let me know what you think @alamb @jackwener

**Comment by jackwener** [comment]
I think put `analyzer rule` in optimizer isn't a good idea.
Because they are two part, and I think someday we may put analyzer into a independent part.

**Review by alamb — COMMENTED** [review]
Thanks @waynexia  -- this looks like a good idea to me. 

Could you also add a "end user" example of how to use this new analyzer rule framework to datafusion-examples?

Perhaps we can extend the existing one here:

https://github.com/apache/arrow-datafusion/blob/a6dcc2db250ee746a3b19c52c90337f74b020c20/datafusion-examples/examples/rewrite_expr.rs#L49-L56

The rationale is both to document the API for potential users but also to help ensure that we don't accidentally remove / break this API in the future during a refactpor

**Comment by waynexia** [comment]
>The rationale is both to document the API for potential users but also to help ensure that we don't accidentally remove / break this API in the future during a refactpor

Great suggestion! 👍 I have expanded the example to illustrate how to create and implement a basic analyzer rule.

File: datafusion-examples/examples/rewrite_expr.rs [hunk 2]
```diff
@@ -66,11 +75,61 @@ fn observe(plan: &LogicalPlan, rule: &dyn OptimizerRule) {
...
+}
+
+impl MyAnalyzerRule {
+    fn analyze_internal(plan: &LogicalPlan) -> Result<LogicalPlan> {
+        // optimize child plans first
+        let mut new_inputs = plan
+            .inputs()
+            .iter()
+            .map(|p| Self::analyze_internal(p))
+            .collect::<Result<Vec<_>>>()?;
```
Comment (jackwener) [inline datafusion-examples/examples/rewrite_expr.rs:98]: we can use `transform_up`

**Review by alamb — APPROVED** [review]
Looks great to me -- thank you @waynexia  and @jackwener

**Review by jackwener — APPROVED** [review]
Thanks @waynexia

**Pull request merged**